src/PKSim.Infrastructure/ORM/Repositories/RemoteTemplateRepository.cs

```diff
@@ -0,0 +1,126 @@
+using PKSim.Core.Services;
+using PKSim.Core.Snapshots.Services;
+
+namespace PKSim.Infrastructure.ORM.Repositories
+{
+   public class RemoteTemplateRepository : StartableRepository<RemoteTemplate>, IRemoteTemplateRepository
+   {
+      private readonly IPKSimConfiguration _configuration;
+      private readonly IJsonSerializer _jsonSerializer;
+      private readonly ISnapshotTask _snapshotTask;
+      private readonly List<RemoteTemplate> _allTemplates = new List<RemoteTemplate>();
+      public string Version { get; private set; }
+
+      public RemoteTemplateRepository(
+         IPKSimConfiguration configuration,
+         IJsonSerializer jsonSerializer,
+         ISnapshotTask snapshotTask)
+      {
+         _configuration = configuration;
+         _jsonSerializer = jsonSerializer;
+         _snapshotTask = snapshotTask;
+      }
+
+      public IReadOnlyList<RemoteTemplate> AllTemplatesFor(TemplateType templateType)
+      {
+         Start();
+         return _allTemplates.Where(x => x.Type.Is(templateType)).ToList();
+      }
+
+      public RemoteTemplate TemplateBy(TemplateType templateType, string name) => AllTemplatesFor(templateType).FindByName(name);
+
+      public async Task<T> LoadTemplateAsync<T>(RemoteTemplate remoteTemplate)
+      {
+         var localFile = Path.Combine(_configuration.RemoteTemplateFolderPath, fileNameWithVersionFor(remoteTemplate));
+         if (!FileHelper.FileExists(localFile))
+            await downloadRemoteFile(remoteTemplate.Url, localFile);
+
+         var buildingBlockType = EnumHelper.ParseValue<PKSimBuildingBlockType>(remoteTemplate.Type.ToString());
+         return await _snapshotTask.LoadModelFromProjectFileAsync<T>(localFile, buildingBlockType, remoteTemplate.Name);
+      }
+
+      public IReadOnlyList<RemoteTemplate> AllReferenceTemplatesFor<T>(RemoteTemplate remoteTemplate, T loadedTemplate)
+      {
+         //We only have reference templates for very specific building block types. So we filter for those
+         switch (loadedTemplate)
+         {
+            case Compound compound:
+               return metabolitesFor(compound);
+            case Individual individual:
+               return expressionProfileFor(remoteTemplate, individual);
+            default:
+               return Array.Empty<RemoteTemplate>();
+         }
+      }
+
+      private IReadOnlyList<RemoteTemplate> expressionProfileFor(RemoteTemplate remoteTemplate, Individual individual)
+      {
+         //TODO Not implemented yet. It will be done with the profile defined for individual as separate building block
+         return Array.Empty<RemoteTemplate>();
+      }
+
+      private IReadOnlyList<RemoteTemplate> metabolitesFor(Compound compound)
+      {
+         return compound.AllProcesses<EnzymaticProcess>()
+            .Select(x => x.MetaboliteName)
+            .Where(x => x.StringIsNotEmpty())
+            .Distinct()
+            .Select(meta => TemplateBy(TemplateType.Compound, meta))
```

Comment:
@Yuri05 Here I am assuming that the reference are also remote templates. They can be in different files. Since they are found by name, we cannot have two templates with the same name. Or we restrict it to being in the same file... I think #2 is not ok for Indiviudal and ExpressionProfile. But it could make sense for compound. Thoughts?

Reply:
How are we going to deal with **multiple versions** of referenced BB-template?
E.g. let's say
* we have 3 template versions for Individual "I1": 1.0, 1.1 and 2.0 
* I1 references expression profile "E1" 
* and we have 2 Template versions for "E1": 1.0 and 1.1

Now if we load I1-template in version 1.1 - which version of E1-template should we take? Theoretically, all I1-versions (1.0, 1.1, 2.0) could have been developed with the same E1-version.

So I think for compounds and metabolites (but also for Individuals and Expressions etc.) the best way would be
* try the **same file** first. If the referenced-BB is found - take it. Then it's the "matching" version for sure.
* If not - try remote. If there are multiple versions of remote BB template available - ask user to select one? Not sure

Or for now we say: always take the same file. Later, with Individuals and Expressions, we might extend it to the 2nd option ("take referenced template from another remote file")

Reply:
>Or for now we say: always take the same file. Later, with Individuals and Expressions, we might extend it to the 2nd option ("take referenced template from another remote file")

Sounds good. We can also say it's the one for compound and the other for individual. Mixing the two will be trickier..

Reply:
Implemented